wysteria/middleware/impl_nats.py

```diff
@@ -85,12 +59,68 @@ def retry_func(*args, **kwargs):
         for count in range(0, NATS_MSG_RETRIES + 1):
             try:
                 return func(*args, **kwargs)
-            except Exception as e:
+            except (RequestTimeoutError, Empty) as e:
                 if count >= NATS_MSG_RETRIES:
                     raise
     return retry_func
 
 
+class _TornadoNats(threading.Thread):
+    """
+    Tiny class to handle queuing requests through tornado.
+    """
+    _MAX_RECONNECTS = 10
+
+    def __init__(self, url):
+        threading.Thread.__init__(self)
+        self._conn = None
+        self._url = url
+        self._outgoing = Queue()
+        self._running = False
+
+    @tornado.gen.coroutine
+    def main(self):
+        """Connect to remote host(s)
+
+        Raises:
+            Exception if unable to establish connection to remote host(s)
+        """
+        self._conn = Client()
+        yield self._conn.connect(
+            servers=[self._url],
+            allow_reconnect=True,
+            max_reconnect_attempts=self._MAX_RECONNECTS
+        )
+        while self._running:
+            while not self._outgoing.empty():
```

Comment:
Is this going to spin the CPU when there are no pending requests? Maybe a better way would be to remove the empty check, let the queue.get() block, and send it a poison pill when you close the connection. Send it a None when you close, and then check for None here and bail.

Reply:
Good idea!